만국박람회[STEP2] Brad, Groot

bradheo65 opened the pull request:
안녕하세요 @ryan-son Step2 PR 보내 드립니다.
생각보다 고민할 거리가 많아져서 늦었네요..😵‍💫😵‍💫

이번에도 잘 부탁드립니다!!

# Step2 PR

## 📑 배경

Json 데이터를 이용해서 만국박람회 소개와 한국의 출품작을 소개하는 앱을 구현

## 🧑🏻‍💻 작업 내용

### 1. **ExpoData**

```
exposition_universelle_1900.json 파일을 Decode 하기 위한 타입 구현.
```

### 2. **ExpoMainViewController**
```
메인 ViewController 구현.
```

### 3. **KoreaItemTableViewController**
```
한국의 출품작 ViewController 구현.
```

### 4. **KoreaItemTableViewCell**

```
한국의 출품작 TableView Cell 타입 구현

```

### 5. **ItemDetailViewController**
```
한국의 출품작 상세 설명 ScrollView 타입 구현
```

### 6. **SendDataDelegate**
```
Delegation방식의 Data 전달을 위한 Protocol 구현
```

### 7. **Int**

```
Int 타입 Data parsing을 위한 Extention 구현
```

### 구현내용

앱 실행 예시



|앱 실행 예시|
|:---:|
|<image src = "https://user-images.githubusercontent.com/45350356/174560674-70a74530-f471-4e81-b90f-b5d5b5c7cd6f.gif" width="300" height="550">
|


## 💡 테스트 방법

앱을 실행하여 확인

## ✏️ 리뷰 노트

### 1. 코드 네이밍 시 보통 단수, 복수형으로 작성하는지 궁금 합니다.

- 예를 들어 `TableView`의 `cell` 안에 item이 여러개 있는 상황이라면 `Items`, 그렇지 않으면 `item`으로 작성해줘야 코드만 보았을 때 그 뜻이 명확한 지 헷갈리네요!

### 2. `sceneDelegate`와 `appDelegate`의 경우는 Controller로 분리해야 할 지 View로 분리해야 할 지 고민을했습니다.

- MVC에 맞게 파일을 분리하는 과정에서 `sceneDelegate`와 `appDelegate`를 어떤 영역으로 봐야하는지에 대한 고민을 했습니다. `sceneDelegate`의 경우 View Life Cycle을 담당하고 appDelegate는 App Life Cycle의 역할을 한다고 알고 있습니다. 각각의 역할에 맞게 분리한다면 어떤 영역으로 봐야 하는지에 대해 생각해 보았습니다. 라이언은 어떻게 생각하시는지 궁금합니다.
    - 브래드: `sceneDelegate`와 `appDelegate`는 화면 전환에 대한 역활을 해주는 것으로 생각하고 있고, `UIKit`를 라이브러리로 가지고 가는 기능들은 `View`로 생각해서 Controller로 생각했습니다.
    - 그루트: `sceneDelegate`에서 Window를 구성하고, 직접적으로 View를 조작하는게 아닌 View의 변화를 전달받아 특정 메서드를 실행하기 때문에 View의 영역이라고 생각했습니다. `appDelegate`는 사실 잘 모르겠습니다.'

bradheo65 committed:
feat: Create Expo1900 ScrollView

- create expo1900ScrollView property
- create setScrollViewConstraints function

Groot-94 committed:
feat: Create Expo1900 StackView, Labels, ImageView

- create expo1900StackView property
- create Labels(titleLabel, visitorsLabel, locationLabel, durationLabel, descriptionLabel)
- create ImageView(posterImageView)

bradheo65 committed:
feat: Create Expo1900 setStackViewConstraints function

Groot-94 committed:
feat: Create Expo1900 decodingJsonData function

bradheo65 committed:
feat: Create ExpoData Type(Model of Data Maneger)

- create Each Json Value Property
- create decodingJsonData Function

Groot-94 committed:
refactor: Remove decodingJsonData function of Expo1900ViewController

Groot-94 committed:
feat: Create Int extension(numberFormatter function)

- Create numberFormatterParse function

bradheo65 committed:
feat: Put poster image to Expo1900StackView

bradheo65 committed:
refactor: Change options of Labels, StackView

- Change alignment, spacing of StackView
- Remove default text of all Labels
- Change font, numberOfLines, textAlignment of titleLabel
- Change numberOfLines of descriptionLabel

Groot-94 committed:
feat: Create Expo1900 subStackView, nextViewButton, FlagImageViews

- Create expo1900SubStackView property
- Create nextViewButton property
- Create FlagImageViews(leftFlagImageView, rightFlagImageView)
- Create setSubStackViewConstraints function

bradheo65 committed:
feat: Create nextViewButton Action of Change View

- Expo1900ViewController -> KoreaItemsTableViewController
--- add Task ---
- Set Navigation ViewController in SceneDelegate.swift
- Remove Main.storyboard, Info.plist

bradheo65 committed:
docs: Add GroundRole.md

bradheo65 committed:
docs: Update Readme.md

bradheo65 committed:
Update README.md

bradheo65 committed:
Update README.md

Groot-94 committed:
feat: Set backgroundColor and hide NavigationBar of Expo1900 View

- Set backgroundColor systemBackground

bradheo65 committed:
Update README.md

Groot-94 committed:
Merge branch 'Step02' of https://github.com/bradheo65/ios-exposition-universelle into Step02

# Conflicts:
#	README.md

bradheo65 committed:
refactor: Encapsulation methods of Expo1900ViewController

- setLabelText
- setViewConstraints
- addUIItemStackView

Groot-94 committed:
feat: Create KoreaItemsTableViewCell

- Create koreaItemsStackView, koreaItemsSubStackView
- Create itemImageView, titleLabel, shortDescriptionLabel
- Create initialize
- Create setConstraints for Cell AutoLayout

bradheo65 committed:
feat: Create Property of KoreaItems and Decoding items Method

Groot-94 committed:
feat: Create koreaItemsTableView

- Set koreaItemsTableViewController tableView to koreaItemsTableView

Groot-94 committed:
feat: Set KoreaItemsTableViewCell

- Put data of KoreaItemsTableViewCell Items

bradheo65 committed:
refeat: Modify navigation setting of each ViewController

- each ViewController(Expo1900ViewController, KoreaItemsTableViewController)

Groot-94 committed:
feat: Create TableViewCell click event and ItemDetailViewController

bradheo65 committed:
feat: Using the Delegate Pattern to Send Data Between Views

Groot-94 committed:
feat: Create itemDetailScrollView

- Create setScrollViewConstraints method for AutoLayout

bradheo65 committed:
feat: Create itemDetailStackView

- Create setStackViewConstraints method for AutoLayout

Groot-94 committed:
feat: Create itemImageView, descriptionLabel

- Create addItemStackView method to add subviews in itemDetailStackView
- Create setKoreaItemData method to insert data

Groot-94 committed:
refactor: Modify code convention of ExpoMainViewController

- Modify name (ExpoMainViewController, mainStackView, subStackView)
- Modify leadingAnchor, trailingAnchor of setStackViewConstraints method
- Modify numberOfLines of  titleLabel (2 -> 0)
- Modify Declaration of Global variable -> Local variable(ExpoData)

bradheo65 committed:
refactor: Modify code convention of KoreaItemTableViewController

- Modify name (KoreaItemTableViewController, koreaItems, koreaItemsTableView)
- Move SendDataDelegate protocol (KoreaItemTableViewController -> Model/Protocol/SendDataDelegate)

Groot-94 committed:
refactor: Modify code convention of KoreaItemTableViewCell

- Modify name (KoreaItemTableViewCell, itemStackView, itemSubStackView)
- Add heightAnchor of itemImageView
- Separate setConstraints method (addUIItemStackView, setViewConstraints)

bradheo65 committed:
refactor: Modify code convention of ItemDetailViewController

- Modify name (Item, itemScrollView, itemStackView)
- Add heightAnchor of itemImageView

Groot-94 committed:
chore: Modify file placement, remove Main.storyboard

Groot-94 committed:
refactor: Bundle duplicate code into method of ExpoMainViewController

- Create createImageView method
- Create createLabel method

bradheo65 committed:
docs: Update Readme, Step01, Step02

ryan-son commented on Docs/GroundRule.md:
```diff
@@ -0,0 +1,64 @@
+
+- [Swift Statements](https://docs.swift.org/swift-book/ReferenceManual/Statements.html) 형식 준수
+- 함수안에 있는 코드 다 붙이기
+- 클래스 100줄, 함수 10줄 지키기
```
라인수 지키기는 쉽지 않은 목표네요. 제 생각에는 기계적으로 라인수를 맞추기보다 타입에 정의된 개별 요소가 이해하기 좋다면 타입의 라인수가 다소 길어지더라도 괜찮다고 생각해요. 다만 해당 역할을 해당 타입이 수행하는 것이 맞는지, 별도로 분리할 필요가 있을지에 대한 고민이 선행되어야 할 것 같네요. 자신의 역할을 벗어난 기능을 제공하기보다 해당 역할을 전문적으로 수행해주는 타입으로 책임을 분리하면 이 고민이 조금은 해결될 수 있을 것 같네요. 

먼저 이야기한 **개별 요소가 이해하기 좋다면**이라는 것의 의미는 이니셜라이저나 메서드, 프로퍼티의 구현부가 간단명료한지를 의미해요. 이 조건을 만족하는지 알아보려면 메서드의 경우 이름으로써 자신이 수행하는 기능을 완벽히 설명하고 있는지를 살펴보는 것이 예시가될 수 있겠네요.

bradheo65 replied:
음 네!! 좋은 말씀인것같아요
기계적으로 라인수에 맞춰서 작성하는 것 보단 개별 요소에 대해 간단명료하게 작성하는게 좋을 것 같습니다

ryan-son commented on Docs/GroundRule.md:
```diff
@@ -0,0 +1,64 @@
+- [Swift Statements](https://docs.swift.org/swift-book/ReferenceManual/Statements.html) 형식 준수
+- 함수안에 있는 코드 다 붙이기
+- 클래스 100줄, 함수 10줄 지키기
+- extention, Mark Comment를 이용한 코드분리
```
코드 분리는 어떤 기준으로 수행하기로 하셨나요~?

Groot-94 replied:
크게는 프로퍼티와 메서드, Life Cycle 세 가지로 나누고 그 안에서 뷰를 조작하는 부분과 그렇지 않은 부분으로 나눠 봤습니다.

ryan-son commented on Docs/GroundRule.md:
```diff
@@ -0,0 +1,64 @@
+- 함수안에 있는 코드 다 붙이기
+- 클래스 100줄, 함수 10줄 지키기
+- extention, Mark Comment를 이용한 코드분리
+- 매직넘버, 매직스트링 X
```
살펴보겠습니다 🤓

ryan-son commented on Expo1900/Expo1900/Model/ExpoData.swift:
```diff
@@ -0,0 +1,29 @@
+import UIKit.NSDataAsset
+
+struct ExpoData {
+    private(set) var title: String?
+    private(set) var visitors: String?
+    private(set) var location: String?
+    private(set) var duration: String?
+    private(set) var description: String?
```
프로퍼티를 모두 옵셔널 타입으로 설정하신 이유가 있을까요? JSON 값 중에 이 요소들이 빠질 수 있다고 생각하셨던 걸까요?

Groot-94 replied:
ExpoData 인스턴스 생성 시 필요하지 않는 값을 정해주는 방법보다 decodingJsonData 함수를 수행하고 나서 값이 들어가는 게 맞는다고 생각해서 옵셔널로 처리했습니다.

ryan-son commented on Expo1900/Expo1900/Model/ExpoData.swift:
```diff
@@ -0,0 +1,29 @@
+        location = "개최지 : \(expoInformation.location)"
+        duration = "개최기간 : \(expoInformation.duration)"
+        description = expoInformation.description
+    }
```
JSON에서 모델로 변환하는 기능이니 모델이 이 역할을 수행하기보다 외부에서 작업을 수행하는게 좋아보이네요. 읽어들인 데이터에서 포매팅이 필요하다면 새로운 인스턴스를 만들어주면 될 것 같아요.

```swift
// Namespace for avoiding literal strings
enum Asset {
    static let expoInformation: String = "exposition_universelle_1900"
}

private func fetchExpoInformation(from assetName: String) -> ExpoInformation? {
    guard let expoInformationAsset = NSDataAsset(name: assetName),
          let expoInformation = try? JSONDecoder().decode(ExpoInformation.self, from: expoInformationAsset.data) else {
        return nil // or throw error
    }
    return expoInformation
}

extension ExpoInformation {
    func toDomain() -> ExpoData {
        return ExpoData(
            title: "방문객 : \(visitors.numberFormatterParse()) 명",
            visitors: "개최지 : \(location)",
            duration: "개최기간 : \(duration)",
            description: description
        )
    }
}

// how to use
guard let expoInformation = fetchExpoInformation(from: Asset.expoInformation) else { return } // or handle error
let expoData = expoInformation.toDomain()

Groot-94 replied:
수정해봤습니다.
[ab377a1](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/ab377a1b34d791a56b32a940dd9bdf85deabf0b4)
매직 리터럴 처리를 깜빡하고 안했네요.. extension으로 포매팅을 나눠주는 방법은 상상도 못했습니다! 상세한 예시 감사합니다.

ryan-son commented on Expo1900/Expo1900/Model/Extention/Int.swift:
```diff
@@ -0,0 +1,19 @@
+import Foundation
+
+extension Int {
+    func numberFormatterParse() -> String {
```
정수 타입에 포매팅이라는 문구면 어떤 기능을 수행할지 예상 가능할 것 같아요. `formatted()`, `formatted(style: .decimal)`, `decimalStyleFormatted()`와 같이 구성해보는건 어떨까요?

Groot-94 replied:
확장성을 고려해서 formatted(style: .decimal) 방법을 채택했습니다.
[de7d54b](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/de7d54b8e736858cb576501f721d591cfe9b9983)

ryan-son commented on Expo1900/Expo1900/Model/Extention/Int.swift:
```diff
@@ -0,0 +1,19 @@
+        guard let result = numberFormatter.string(from: self as NSNumber) else {
+            return ""
+        }
+        return result
```
이런건 보통 선호하는 스타일이 제각각이지만 `result`가 결국 반환값을 의미한다면 저는 아래처럼 작성해보는 것도 고려해볼 것 같아요. 완전히 취향의 영역이니 원하시는대로 하시면 됩니다. 옵셔널 바인딩을 통해서 예외처리를 폭넓게 할 수 있으니 웬만하면 옵셔널 바인딩을 사용하는 것을 권장드려요.

```swift
return numberFormatter.string(from: self as NSNumber) ?? ""

ryan-son commented on Expo1900/Expo1900/Model/Protocol/SendDataDelegate.swift:
```diff
@@ -0,0 +1,10 @@
+
+protocol SendDataDelegate: AnyObject {
+    func sendItemData(item: KoreaItem)
+}
```
이 메서드가 수행하는 것은 사실 `send`보다는 `set`에 가까우니 다른 이름을 사용해보는건 어떨까요? 저는 `KoreaItemSettable`이라는 이름을 생각해봤어요.

Groot-94 replied:
공식문서에 Protocol naming 방법이 따로 있음을 알 수 있었습니다... 기본을 놓치고 있었습니다. 감사합니다! 함수 이름도 변경해봤습니다.

[77a6127](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/77a6127edbb36e7d2b28855b9f1b11f1bc8a125d)

ryan-son commented on Expo1900/Expo1900/View/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,85 @@
+class KoreaItemTableViewCell: UITableViewCell {
+    //MARK: - KoreaItem Cell View
+    
+    let itemStackView: UIStackView = {
```
외부에서 활용하지 않는다면 접근제어를 걸어줄 수 있을 것 같네요~

Groot-94 replied:
수정했습니다!
[d8ec638](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/d8ec638ea507ae311e1d31ae6b411fc97fb458a2)

ryan-son commented on Expo1900/Expo1900/View/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,85 @@
+    let itemImageView: UIImageView = {
+        let imageView = UIImageView()
+        imageView.contentMode = .scaleAspectFit
+        imageView.heightAnchor.constraint(lessThanOrEqualToConstant: 80).isActive = true
```
사실 코드는 어느 위치에 있어도 괜찮지만, 저는 관련있는 코드는 모아두는 편이에요. 여러분 선호에 따라 작성하시면 됩니다~

Groot-94 replied:
constraint를 정하는 부분들을 함수로 모아봤습니다. 

[d8ec638](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/d8ec638ea507ae311e1d31ae6b411fc97fb458a2)

ryan-son commented on Expo1900/Expo1900/View/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,85 @@
+        setViewConstraints()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
```
이 이니셜라이저는 언제 호출되나요? 이 메서드가 호출될 가능성은 존재하지 않나요?

Groot-94 replied:
>required init?(coder: NSCoder)에 대해 찾아봤을때 UIView 가 NSCoding protocol을 채택하고 그 서브클래스에서 init을 호출하면서 필수로 해줘야 하는 init이라고 알 수 있었습니다. 

>required init?(coder: NSCoder호출되는 시점은 main storyboard 에 등록?(변환?)할 때 호출된다고 찾아봤습니다.. 
fatalError 메서드는 init이 실패하면 에러처리를 위해 호출된다고 알고 있습니다.

>NSCoder의 개념은 커스텀 셀의 코드를  Interface builder에서 사용하는 언어로 변경해주기 위한 타입이라고 생각을 해봤습니다.

**결론 : 저희는 UI를 코드로 구현했기 때문에 required init?(coder: NSCoder)이 호출되지 않는다고 생각했습니다. 그렇기 때문에 메서드도 가능성이 존재하지 않는다고 생각했습니다.**

저희가 생각한게 맞을까요... 상당히 어려운 개념이라서 확실하게 이해했는지 확신이 없습니다..

bradheo65 replied:
추가적으로 UIView와 UIViewController는 NSCoding 프로토콜을 구현하고 있으므로 상속받은 클래스에서 `requied init()`을 해줘야 되지 않을까 생각 되네요

ryan-son replied:
스토리보드나 xib를 통해 뷰를 생성하면 인스턴스가 이 이니셜라이저로 생성됩니다. 말씀하신 것처럼 `UIView`와 `UIViewController`는 `NSCoding`을 준수하므로 [`NSCoder`](https://developer.apple.com/documentation/foundation/nscoder)를 통해 객체가 아카이브 될 수 있습니다. 아카이브 된다는 것은 디스크에 저장이 된다는 의미에요. 그래서 이 이니셜라이저가 호출되면 여러분이 스토리보드나 xib를 통해 복잡하게 구성해두었던 뷰의 속성, 제약, Interface builder outlet, action들이 언어카이빙 (로드)되어 다시 본래의 객체 그래프를 복원되는 것입니다. 브래드가 말씀하신 것처럼 상속하여 구현해주어도 좋고, 스토리보드와 xib를 사용하지 않을 것이라면 힘들이지 않고 구현하셔도 괜찮습니다.

ryan-son commented on Expo1900/Expo1900/View/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,85 @@
+    
+    //MARK: - Setting View Methods
+    
+    private func addUIItemStackView() {
```
`UIKit`의 요소에는 대부분 `UI`라는 접두어가 붙으니 생략할 수 있을 것 같아요.

Groot-94 replied:
생각해보니 굳이 쓰지 않아도 되는 느낌이 들어서 제거했습니다.
[d8ec638](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/d8ec638ea507ae311e1d31ae6b411fc97fb458a2)

ryan-son commented on Expo1900/Expo1900/View/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,85 @@
+        itemStackView.bottomAnchor.constraint(equalTo: self.contentView.bottomAnchor).isActive = true
+        itemStackView.leadingAnchor.constraint(equalTo: self.contentView.leadingAnchor).isActive = true
+        itemStackView.trailingAnchor.constraint(equalTo: self.contentView.trailingAnchor).isActive = true
+        itemSubStackView.widthAnchor.constraint(equalTo: self.itemImageView.widthAnchor, multiplier: 4).isActive = true
```
매직넘버라고 볼 수 있을까요? `ViewController` 안에 `Metric`이라는 타입을 두고 이러한 크기 요소를 따로 관리해도 좋을 것 같아요.

Groot-94 replied:
매직넘버와 매직리터럴의 정의를 코드에서 숫자나 문자로 변경이 가능한 부분이라고 생각해서 조금 극단적으로 모두 바꾸는 방법을 사용해봤습니다. 매직넘버와 매직리터럴의 기준을 팀과 상의해서 정해야 하는지 아니면 명확하게 정해져있는지 궁금하네요.. 

[d8ec638](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/d8ec638ea507ae311e1d31ae6b411fc97fb458a2)

ryan-son replied:
명확한 것은 없어요. 두 번 이상 사용되면 정의해야 한다는 기준을 가진 사람도 있고, 한 번 사용되더라도 코드에서 리터럴 값은 없는 것이 좋다는 의견도 있죠. 팀에서 기준을 정하면 됩니다.

ryan-son commented on Expo1900/Expo1900/Controller/KoreaItemTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+
+import UIKit
+
+class KoreaItemTableViewController: UITableViewController {
```
`UIViewController`에 `UITableViewDataSource`, `UITableViewDelegate`를 채택하여 구현하지 않고, `UITableViewController`를 상속하여 구현하신 이유가 있을까요? 앞서 언급한 방식과 이렇게 구현한 방식 간에는 어떤 차이점이 있었나요?

Groot-94 replied:
두번째 화면의 전체적인 뷰가 테이블 뷰라고 생각해서 사용했습니다. 

UITableViewController는 뷰 자체가 테이블 뷰이기 때문에 뷰를 구성하는 화면전체가 테이블 뷰인 경우에 사용해야 한다고 생각합니다.
그와 반대로 UIViewController에 UITableViewDataSource, UITableViewDelegate를 채택하는 방법은 테이블 뷰를 서브뷰로 가져야 하는 경우에 쓴다고 생각합니다. 저희가 생각하는게 맞을까요..?
 [참고문서](https://developer.apple.com/library/archive/documentation/UserExperience/Conceptual/TableView_iPhone/CreateConfigureTableView/CreateConfigureTableView.html#//apple_ref/doc/uid/TP40007451-CH6-SW10)

ryan-son replied:
정답은 없습니다. 전체적인 뷰가 테이블 뷰가 아니라 일부가 테이블뷰를 가지는 경우에도 `UITableViewController`를 이용할 수 있고, 반대의 경우에도 `UITableViewDataSource`를 준수하여 전체 뷰를 테이블뷰로 구성할 수 있습니다. 두 방식에 차이는 없지만, 저는 `UITableViewDataSource`와 `UITableViewDelegate` 코드를 `extension`을 통해 구조화하는 것을 선호하는 편이라 `UIViewController`에 채택해서 구현하는 방식을 선호하기는 하지만, 정말 아무래도 관계 없으니 변경하실 필요는 없습니다.

ryan-son commented on Expo1900/Expo1900/Controller/KoreaItemTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+class KoreaItemTableViewController: UITableViewController {
+    //MARK: - KoreaItemTable Property
+    
+    var item = [KoreaItem]()
```
1. 외부에서 활용할 요소가 아니라면 숨겨두는 것도 좋겠네요~
2.  `Array`, `Set`, `Dictionary`와 같은 Collection type은 해당 인스턴스가 Collection 내부에 복수로 존재하니 보통 복수형의 이름을 사용합니다. 그러니 이 경우에는 `items`가 되겠네요.
3. `item`이라는 이름을 통해서는 어떤 `item`인지 추측하기 어려우니 `koreaItem`으로 풀어서 표현하시는건 어떨까요?

bradheo65 replied:
넵 라이언! 혼동되는 부분이였는데 명쾌하게 설명주셔서 감사합니다!
[1f94936](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/1f9493674af432d719ef29352d3c081aa8070fc0)

ryan-son commented on Expo1900/Expo1900/Controller/KoreaItemTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+    
+    let itemTableView: UITableView = {
+        let tableView = UITableView()
+        tableView.register(KoreaItemTableViewCell.self, forCellReuseIdentifier: "KoreaItemTableViewCell")
```
구현부에 Literal string을 사용하면 실수하는 경우가 가끔 있더라구요. 아래와 같은 방식을 적용해보는건 어떨까요?

```swift
class KoreaItemTableViewCell: UITableViewCell {
    static var reuseIdentifier: String {
        return String(describing: self)
    }
}
```

여기에서 생각을 확장해보면 protocol default implementation을 통해 cell registration, dequeueReusableCell을 수행할 때도 `reuseIdentifier`를 신경쓰지 않고 작업을 하게 만들 수도 있을거에요.

```swift
tableView.register(KoreaItemTableViewCell.self)
```

ryan-son commented on Expo1900/Expo1900/Controller/KoreaItemTableViewController.swift:
```diff
@@ -0,0 +1,69 @@
+            return
+        }
+        self.item = decodedKoreaItems
+    }
```
이 메서드는 에셋을 가져와서 디코드하고 item에 세팅하는 작업을 수행하고 있네요. 수행하는 일이 많아서 현재의 메서드 이름으로는 수행하는 작업을 제대로 설명하기는 어려울 수 있겠네요. 로직을 그대로 둔다고 하더라도 이름은 디코딩에 집중하기보다 item을 세팅한다에 초점을 맞추는게 좋아 보입니다!

bradheo65 replied:
음 넵 좋은 의견인거 같아요!! 그래서 아래 코드대로 수정해보았습니다.
```swift
    private func fetchExpoInformation(from assetName: String) -> ExpoInformation? {
        guard let expoInformationAsset = NSDataAsset(name: assetName),
              let expoInformation = try? JSONDecoder().decode(ExpoInformation.self, from: expoInformationAsset.data) else {
            return nil
        }
        return expoInformation
    }
```
라이언이 의도하신 부분과 동일할까요~?

ryan-son replied:
좋습니다! 👍 이건 데이터를 로컬에서 가져오는 것이라 실패해도 원인을 금방 찾을 수 있으니 이런 형식의 에러처리도 괜찮을 것 같지만, 이후에 네트워킹을 통해 데이터를 원격에서 가져오는 경우에는 옵셔널값을 반환 받았을 때 원인 분석이 어려워지는 경우가 있으니 이후에는 다양한 에러 처리 방식을 고민해보시면 좋겠습니다~

ryan-son commented on Expo1900/Expo1900/Controller/SceneDelegate.swift:
```diff
@@ -12,10 +12,12 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+        let mainViewController = ExpoMainViewController()
+        let navigationController = UINavigationController(rootViewController: mainViewController)
+        window?.rootViewController = navigationController
+        window?.makeKeyAndVisible()
```
코드 기반으로 UI를 작업하시려고 window를 설정하는 코드를 작성해보셨네요. `makeKeyAndVisible()`은 어떤 역할을 수행하는 메서드였나요?

bradheo65 replied:
공식문서에 창을 표시하고 그것을 `key`창으로 만들어주는 역활을 한다고 나와있네요.

풀어서 생각한다면 준비했으니까 원하는 뷰를 보이게 하라는 의미이지 않을까 생각 됩니다

ryan-son replied:
`AppDelegate` 또는 `SceneDelegate`에서 확인할 수 있는 `UIWindow` 타입의 `window`라는 변수가 있는데 스토리보드로 작업할 때는 이걸 신경쓰지 않아도 됐었던 것을 기억하실겁니다. 하지만 이 작업을 직접 `UIWindow`를 만들어 수행하고 있으니 직접 만든 `UIWindow`가 `keyWindow`, 그리고 보이는 상태로 만들기 위해 사용한다는 것을 공식문서를 통해 확인하셨을 것 같네요~!

ryan-son commented on Expo1900/Expo1900/Controller/ExpoMainViewController.swift:
```diff
@@ -0,0 +1,148 @@
+
+class ExpoMainViewController: UIViewController {
+    //MARK: - Expo Main Property
+    var expoData = ExpoData()
```
사실상 프로퍼티가 모두 옵셔널인 타입의 인스턴스인데, 타입 자체를 옵셔널로 선언하는건 어떨까요?

```swift
private var expoData: ExpoData?
```

bradheo65 replied:
네 그렇네요.. 옵셔널로 선언해두었는데 사실 이런 형태를 처음보아서 새롭게 알게 되었네요!! 감사합니다!

ryan-son commented on Expo1900/Expo1900/Controller/ExpoMainViewController.swift:
```diff
@@ -0,0 +1,148 @@
+        let scrollView = UIScrollView()
+        scrollView.translatesAutoresizingMaskIntoConstraints = false
+        return scrollView
+    }()
```
스크롤뷰 구성하시면서 어려움은 없었나요~?

bradheo65 replied:
"스크롤 뷰의 Layout 을 어떤 것을 기준으로 잡아서 크기를 정해줘야 될지"에 대해 고민이 있었습니다.
contentLayoutGuide과  frameLayoutGuide 설정함에 따라서 스크롤 뷰에 표시되는 정보가 화면상에 벗어난 경우도 있고 짤리는 경우도 있었는데 다시 AutoLayoutGuide를 보면서 서로의 대한 차이점을 이해했습니다.

ryan-son replied:
고민되는 부분이 있으시면 말씀해주세요~ 😉

ryan-son commented on Expo1900/Expo1900/Controller/ExpoMainViewController.swift:
```diff
@@ -0,0 +1,148 @@
+        let button = UIButton()
+        button.setTitle("한국의 출품작 보러가기", for: .normal)
+        button.setTitleColor(UIColor.blue, for: .normal)
+        button.setContentCompressionResistancePriority(UILayoutPriority(751), for: .horizontal)
```
CRCH 개념을 적용해보셨네요! 👏

bradheo65 replied:
아무래도 이미지의 우선순의를 줄여주는 것이 사용자 입장에서 보여지기에 더 좋을 것 같아 적용해 보았어요!

ryan-son commented on Expo1900/Expo1900/Controller/ExpoMainViewController.swift:
```diff
@@ -0,0 +1,148 @@
+        self.navigationItem.title = "메인"
+        self.view.backgroundColor = .systemBackground
+        self.view.addSubview(mainScrollView)
+        self.mainScrollView.addSubview(mainStackView)
```
이 작업들도 아래 setLabelText()와 같이 목적을 가진 메서드 단위로 만들어줄 수 있을 것 같네요. 목적에 따라 여러 메서드를 만들 수도 있겠고요.

bradheo65 replied:
정말 좋은 의견인것같습니다!!
```swift
   private func setViewConstraints() {
        setScrollViewConstraints()
        setStackViewConstraints()
        itemImageView.heightAnchor.constraint(lessThanOrEqualToConstant: DetailSetUp.itemImageViewMaxHeight).isActive = true
    }
```
위의 처럼 목적에 따라 여러 메서드를 만들고 묶어서 한번에 호출해주는 방식으로 해보았습니다.

ryan-son commented on Expo1900/Expo1900/Controller/ExpoMainViewController.swift:
```diff
@@ -0,0 +1,148 @@
+    
+    private func addUIItemStackView() {
+        let storeUIView = [titleLabel, createImageView(imageName: "poster"), createLabel(expoData.visitors),
+                           createLabel(expoData.location), createLabel(expoData.duration), descriptionLabel, subStackView]
```
1. 라인을 나누려다 도리어 읽기 어려울 수 있을 것 같은데, 이렇게 작성해보는건 어떠세요? 그리고 여기에서 store는 어떤 의미인가요?

```swift
let storeUIView = [
    titleLabel,
    createImageView(imageName: "poster"),
    createLabel(expoData.visitors),
    createLabel(expoData.location), 
    createLabel(expoData.duration), 
    descriptionLabel, 
    subStackView
]
```

bradheo65 replied:
"`store`는 저장, 가지고 있다" 라고 생각해서 네이밍을 "stackView에 넣어줄 요소들을 가지고 있다." 지어주었습니다.
다시 생각해보니 네이밍을 좀 더 직관적으로 `stackViewItemsArray`스택뷰의아이템들의배열이라고 다시 작성해보았습니다.

ryan-son replied:
`stackViewItemsArray`를 통해서도 파악할 수 있는 정보는 제한적이니 지역적으로 이용되는 상수인만큼 `views` 혹은 `mainViews`로 지어도 괜찮을 것 같아요. 이후에 `mainStackView`에 들어가는 코드를 읽으면 무엇을 위한 배열인지 이해할 수 있을테니까요~ 그래도 무언가 부족하다면 `메인 스택뷰 대상 뷰들`이라는 의미로 `mainStackViewSubjectViews`를 생각해볼 수 있을 것 같아요.

ryan-son commented on Expo1900/Expo1900/Controller/ItemDetailViewController.swift:
```diff
@@ -0,0 +1,95 @@
+class ItemDetailViewController: UIViewController {
+    //MARK: - ItemDetail Property
+
+    var Item: KoreaItem?
```
😉
```suggestion
    var item: KoreaItem?
```

ryan-son commented on Expo1900/Expo1900/Controller/ItemDetailViewController.swift:
```diff
@@ -0,0 +1,95 @@
+
+import UIKit
+
+class ItemDetailViewController: UIViewController {
```
`KoreaItemDetailViewController`는 어떨까요?

bradheo65 replied:
음 처음 네이밍 시 추 후에 `japenItem`처럼 재사용성에 대해 생각을 해서 위에 처럼 작성했는데요. 다시 생각해보니 결국 코드가 KoreaItem`에 대해서만 item을 받는 로직으로 되어 있기에 어색하다는 느낌을 받아서 수정했습니다.

[2b73683](https://github.com/yagom-academy/ios-exposition-universelle/pull/184/commits/2b73683ca09cb90df3534356bb45bb089caed963)

ryan-son commented on Expo1900/Expo1900/Controller/ItemDetailViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    
+    //MARK: -  ItemDetail View
+
+    let itemScrollView: UIScrollView = {
```
접근제한자는 전체적으로 검토가 필요해보이네요~

bradheo65 replied:
넵 리펙토링 하면서 접근제한자 설정해보았습니다.

ryan-son reviewed (changes requested):
그루트, 브래드! Step 2 수행하시느라 고생 많으셨어요. 작성하신 커밋들만 확인해봐도 얼마나 고민하셨는지 확인할 수 있었습니다.

# 질의응답
> - 코드 네이밍 시 보통 단수, 복수형으로 작성하는지 궁금 합니다. 예를 들어 `TableView`의 `cell` 안에 item이 여러개 있는 상황이라면 `Items`, 그렇지 않으면 `item`으로 작성해줘야 코드만 보았을 때 그 뜻이 명확한 지 헷갈리네요!

상황에 따라 단/복수 개념이 바뀌지는 않는다고 생각해요. 복수형 이름은 주로 `Array`, `Set`, `Dictionary`와 같은 `Collection` 타입의 프로퍼티에 붙입니다. 해당 타입의 인스턴스가 컬렉션 안에 복수 존재할 수 있다는 것을 암시하는거죠.

>- `sceneDelegate`와 `appDelegate`의 경우는 Controller로 분리해야 할 지 View로 분리해야 할 지 고민했습니다.
> - MVC에 맞게 파일을 분리하는 과정에서 `sceneDelegate`와 `appDelegate`를 어떤 영역으로 봐야하는지에 대한 고민을 했습니다. `sceneDelegate`의 경우 View Life Cycle을 담당하고 appDelegate는 App Life Cycle의 역할을 한다고 알고 있습니다. 각각의 역할에 맞게 분리한다면 어떤 영역으로 봐야 하는지에 대해 생각해 보았습니다. 라이언은 어떻게 생각하시는지 궁금합니다.

두 요소는 `UIViewController`와 같은 `Controller`로 폴더 분류를 하기에는 성향이 맞지 않는다고 생각해요. `AppDelegate` 같은 경우 `@main` attribute를 통해 알 수 있듯이 앱의 진입점이 되는 중요한 타입이고, `SceneDelegate`의 경우에도 앱이 active, inactive 상태를 감지하여 알맞은 작업을 수행하는 Scene Based Life Cycle을 다루니까요. 그래서 분류를 하는 경우에도 프로젝트 최상위 레벨 폴더에 `App`, `Application`, `Entry`로 분류하는 경우가 많은 것 같아요. 향후에 기능에 따라 별도의 프로젝트로 모듈화를 진행할 경우에도 `AppDelegate`나 `SceneDelegate`는 앱 타겟에 반드시 포함되어야 하니 MVC 분류에 편입하는 것은 맞지 않아 보이네요.

도움이 필요하시거나 궁금한 점이 있다면 DM 부탁드릴게요~!

Groot-94 committed:
refactor: Modify overall of ExpoData

- Separate decodingJsonData (toDomain, fetchExpoInformation)
- Move toDomain, fetchExpoInformation method
- Set Asset for Assets namespace

bradheo65 committed:
refactor: Modify magic string to name space and Method name

Groot-94 committed:
refactor: Modify to final class, access control, magic number(CGFloat)

bradheo65 committed:
refactor: Modify to final class, naming, access control, magic string

- modify item -> koreaItems
- modify decodingJsonData() -> setKoreaItems()

Groot-94 committed:
refactor: Modify to final class, naming, access control, magic string

- Modify storeUIView -> mainStackViewItemsArray
- Modify subStoreUIView -> subStackViewItemsArray
- Set setMainView method for view setting

bradheo65 committed:
refactor: Modify to final class, naming, access control, magic string

- Modify Item -> koreaItems
- Modify storeUIView -> stackViewItemsArray
- Set setMainView method for view setting

Groot-94 committed:
refactor: Modify name of KoreaItemSettable protocol and method

ryan-son commented on Expo1900/Expo1900/Controller/ExpoMainViewController.swift:
```diff
@@ -0,0 +1,189 @@
+            createLabel(expoData?.location),
+            createLabel(expoData?.duration),
+            descriptionLabel,
+            subStackView]
```
마지막 단이 더 잘 구분될 수 있게 줄바꿈 해주는 것도 괜찮아 보입니다~!
```suggestion
            subStackView
        ]
```

ryan-son commented on Expo1900/Expo1900/Controller/ExpoMainViewController.swift:
```diff
@@ -0,0 +1,189 @@
+            createLabel(expoData?.duration),
+            descriptionLabel,
+            subStackView]
+        mainStackViewItemsArray.forEach {self.mainStackView.addArrangedSubview($0)}
```
`{ return }`과 같이 작성하셨던 것처럼 중괄호 컨텐츠는 한 칸씩 띄우는 것이 일반적인 컨벤션이에요.
```suggestion
        mainStackViewItemsArray.forEach { self.mainStackView.addArrangedSubview($0) }
```

ryan-son commented on Expo1900/Expo1900/View/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,104 @@
+    //MARK: - Setting Identifier
+    
+    static var identifier: String {
+        return "KoreaItemTableViewCell"
```
Type safe한 방식을 선호하신다면 `String(describing: KoreaItemTableViewCell.self)`과 같은 방법도 생각해볼 수 있어요.

ryan-son commented on Expo1900/Expo1900/View/KoreaItemTableViewCell.swift:
```diff
@@ -0,0 +1,104 @@
+        setViewConstraints()
+    }
+    
+    required init?(coder: NSCoder) {
```
아무도 그러지 않겠지만, 직접 호출을 막는 방법으로 [available](https://docs.swift.org/swift-book/ReferenceManual/Attributes.html#ID348) attribute를 이용하는 방법이 있어요. 메서드 사용자에게 메서드 이름이 바뀌었음을 알려주거나 메서드가 사용될 수 있는 플랫폼과 버전을 지정하는 등 다양한 방식으로 활용할 수 있으니 이후에 관심이 생기시면 학습해보세요~!
```swift
@available(*, unavailable)
required init?(coder: NSCoder { ... }
```

ryan-son reviewed (approved):
브래드, 그루트, 코멘트가 늦어져서 미안합니다!

Step 2 요구사항을 훌륭하게 작업해주셔서 머지하도록 하겠습니다.
계속해서 화이팅하시고, 막히시거나 궁금한 부분 있으시면 말씀해주세요~! 😊
